### PR TITLE
Security airlocks

### PR DESCRIPTION
**Pull request by Mysak0CZ**
## Idea itself ##
- My main inspiration: [Paradise hacking](http://nanotrasen.se/wiki/index.php/Hacking#Secure_Airlocks)
- Airlocks are now (at least for me) only AIR locks, i can get into vault under 10 seconds...
- Main feature is make door hacking harder, but not harder than destroying doors (there integrity buff comes in)
- However it is supposed to stay unshielded if someone removes protection and than closes panel
- It is also supposed to be easier to hack doors, than get in through walls
## Current features ##
- All airlocks now have var security_level: it goes from 0 to 6 (0-none, 1-simple, 6 full, 2-5 help for construction)
- Level above 1 multiplies door integrity by 1.5
- If security level is above 0, wires cannot be accessed until you remove protection
- Security level do not affect panel_open, and while you are ~~removins security from~~ upgrading doors, you can anytime use screwdriver, to hide what you are doing
- Security level does not affect anything else than integrity and wires (and bombtrap placing)
- If panel is open, tools (if current) are used for construction instead for normal actions
that means if you use welder on opened, secured panel, you will slice throught armor
but if panel is closed, you will normally weld doors shut (even if in middle of deconstruction)
- What tool use next can be easyily guessed by examinating
- if doors have security level, with panel closed, examination will tell, that they are stronger than normal doors
### Construction ###
- If you open wire panel, and it have no security level, it is no difference from pervious version
- in that case you can add 2 metal sheets to upgrade to security 1
- or 2 plasteel sheets to upgrade to security 6
### Deconstruction ###
- To access wires, you need to remove security level
- lvl 1 (metal) can by removed by simply slicing it with welder (and you get back your 2 metal plates)
- lvl 6 can be deconstructed:
1. wirecutters (if doors have power, they will [60%] shock you) (to lvl 5)
2. welder (to lvl 4)
3. crowbar (first plasteel refunded here) (to lvl 3)
4. welder (to lvl 2)
5. crowbar (second plasteel refunded here)  (to lvl 0, yes 1 is reserved for normal metal)

Please note, that during construction you cannot use screwdriver, that closes panel, but wrech can be used. 
Also please note, that maps will need to be edited, to have secure doors at roundstart

# Your ideas #
- What i can easily edit:
added health to airlocks (if any...)
construction/deconstruction proces
material costs / refunds
edit or remove step delays
on what level of security are wires still accessable and when not
- Harder but possible to do for me
new sprites
almost anything you request and others approve

I tesed my code compleatly, but not all sprites (so it is possible, there is misstype in sprites name or missed pixed), if you see something like that while testing this, please tell me so i can fix it

:cl: Mysak0CZ
add: Door can now have higher security, making them stronger and wires harder to access
add: More info can be found on github or wiki (if this passes)
imageadd: Protected wires now have sprites
/:cl:

Last edit: 11:00 UTC 13/11/2016

### CODE REVIEW

**Comment by lzimann** [comment]
\>If security level is above 0, wires cannot be accessed

What

**Comment by Niknakflak** [comment]
what

**Comment by Mysak0CZ** [comment]
please read it compleatly before commenting, it means, you need to remove additional cover to access them

File: code/game/machinery/doors/airlock.dm [hunk 1]
```diff
@@ -459,6 +498,28 @@ var/list/airlock_overlays = list()
...
+			if(AIRLOCK_SECURITY_PLASTEEL_I_S)
+				user << "There is cut plasteel"
+			if(AIRLOCK_SECURITY_PLASTEEL_I)
+				user << "There is inner welded plasteel cover"
+			if(AIRLOCK_SECURITY_PLASTEEL_O_S)
+				user << "There is cut plasteel"
+			if(AIRLOCK_SECURITY_PLASTEEL_O)
+				user << "There is welded plasteel cover"
+			if(AIRLOCK_SECURITY_PLASTEEL)
+				user << "There is protective grile over panel"
```
Comment (swindly) [inline code/game/machinery/doors/airlock.dm]: grille
There should also be articles here and in other messages in this switch.

**Comment by GunHog** [comment]
TL;DR: You can upgrade airlocks with metal/plasteel to increase their health and prevent people hacking doors through wires. You have to cut away the metal/plasteel before you can open the panel and reach the wires.

File: code/game/machinery/doors/airlock.dm [hunk 2]
```diff
@@ -878,6 +942,135 @@ var/list/airlock_overlays = list()
...
+					if(!WT.remove_fuel(2, user))
+						return
+					user << "<span class='notice'>You begin cutting wire shielding...</span>"
+					playsound(loc, WT.usesound, 40, 1)
+					if(do_after(user, 40/WT.toolspeed, 1, target = src))
+						if(!panel_open || !WT.isOn())
+							return
+						playsound(loc, WT.usesound, 50, 1)
+						user.visible_message("<span class='notice'>[user] cuts throught \the [src]'s shielding.</span>",
+										"<span class='notice'>You cut throught \the [src]'s shielding.</span>",
```
Comment (swindly) [inline code/game/machinery/doors/airlock.dm]: through

**Comment by lzimann** [comment]
Honestly... sounds awful imo.

**Comment by GunHog** [comment]
I like the idea myself, given the near-global structure nerf, airlocks are now a joke of a barrier, and I feel that should change.

**Comment by Mysak0CZ** [comment]
yes, now i can get into vault within 10 seconds, with this, i will need insulated gloves and it will take longer

**Comment by kevinz000** [comment]
:-1:
Leave door-hacking as is, vault is secured by the security force, not the door being harder to hack. Same with other doors.
Unless, you don't add this to the map at roundstart, and instead make this a feature. If people want to use extra metal/plasteel to reinforce their autismfort, fine, but don't make secure areas even more secure.

**Comment by Mysak0CZ** [comment]
I will not add it to maps, it is feature and unless someone uses metal or plasteel on doors with open panel, it will be like it is not in here

**Comment by PKPenguin321** [comment]
so tl;dr: the step of airlock hacking at the very beginning where you screwdriver it open can now be made into several different steps depending on the security level of an airlock, many of which require welders.

not a bad idea but the requirement of welders is really lame since it prevents a lot of airlocks from being touched. i'd personally put no delays (do_after()s) into each step so that an experienced player can breeze through these steps, it might already be that way, i haven't checked.

**Comment by Anonmare** [comment]
It's an interesting suggestion. I think it's worth considering, though I'd like to see how it plays ingame before I can give a definitive answer. It'd probably be easier to make people more accepting of the feature if it's something that must be done by players.

**Comment by A-t48** [comment]
Seems pointless unless it's part of the map. Either it becomes a silly metagaming feature, or unused.

**Comment by Shadowlight213** [comment]
I think bay implemented this better via airlock braces. They acted like an extra health level for airlocks and had to be destroyed or removed to open it.
This seems like a huge buff to like any door. I can see it existing for reinforcing airlock heakth, but making them unhackable is too far.

**Comment by XDTM** [comment]
Currently doors ignore damage below 20 and vault doors have a higher deflection value. I'd say to make vault doors start at security 6, make security 6 increase deflection and lower the base value for vault doors to the default 20. Security 1 shouldn't make the door sturdier, just harder to hack.

**Comment by Mysak0CZ** [comment]
@XDTM good idea

File: code/game/machinery/doors/airlock.dm [hunk 3]
```diff
@@ -322,7 +343,10 @@ var/list/airlock_overlays = list()
 			else
 				filling_overlay = get_airlock_overlay("fill_closed", icon)
 			if(panel_open)
-				panel_overlay = get_airlock_overlay("panel_closed", overlays_file)
+				if(security_level)
+					panel_overlay = get_airlock_overlay("panel_closed_protected", overlays_file)
+				else
+					panel_overlay = get_airlock_overlay("panel_closed", overlays_file)
```
Comment (xxalpha) [inline code/game/machinery/doors/airlock.dm]: >editing airlock icon code

**BREATHING INTENSIFIES**

**Comment by Dorsisdwarf** [comment]
I think six security levels is overkill. 

I'd have split it into "no-sec"(maint airlocks/external airlocks, hall airlocks), "Private"(all departmental), and "High-sec" (armory, vault, upload)

**Comment by Mysak0CZ** [comment]
there are six level, but 
0 is none
1 is basic
6 is max
2-5 are deconstruction phases for max security

**Comment by Mysak0CZ** [comment]
In bd9b16b i added, that all doors have normal_integrity
if doors are security level above 1, integrity is multiplied by 1.5 and damage deflection is incresed from 20 to 30
if doors are lvl 1 or 0, integrity and deflection is unaffected
also command, security, high_security and abductor doors starts at level 1
vault and centcomm doors starts at level 6

**Comment by uraniummeltdown** [comment]
Paradise doesn't actually have this in the map, I've hacked into EVA and other secure places just like anywhere else, I've never had to do anything else but screw open the panel and pulse.

**Comment by Mysak0CZ** [comment]
strange

**Comment by Durkel** [comment]
All I can see this being is a pain in the balls. Sybil is having low pop rounds at late night/early morning and It's very common for the head staff to be dead and there to be no AI. The only option we have is to hack the doors to get around the majority of the time and this is just going to make it extremely tedious simply for the sake of being tedious as fuck. That is makes getting a emag even more necessary.

**Comment by xxalpha** [comment]
Nah, this is good. Not all airlocks should be easy to hack.

**Comment by Cyberboss** [comment]
I like this, but ONLY for the the vault and black doors

**Comment by Cyberboss** [comment]
And this better not affect emags

**Comment by Mysak0CZ** [comment]
no, emags are unaffected, but airlock charges need unshielded access to wires

**Comment by XDTM** [comment]
If anyone hasn't read the main post, the covers can still be removed, but take more steps and different tools, they're not just unhackable.

**Comment by GunHog** [comment]
There are not, from a gameplay stance, six 'security' levels, there are three. The normal one, the one where the airlock is plated with extra metal, and one with plasteel. The other levels are simply _deconstruction steps_ for the plasteel airlocks. I will be more than delighted to be able to secure airlocks with this feature.

**Comment by Cyberboss** [comment]
Yeah, you might wanna clarify that in the main post

**Comment by optimumtact** [comment]
I think on the whole this gives more flexibility to mappers.

Mappers should keep in mind that allowing access to high security areas is part of the game.

**Comment by optimumtact** [comment]
Also thanks for this contribution, it was well developed and you laid out your PR idea and theory comprehensively

**Comment by Mysak0CZ** [comment]
Thanks for merging, i am working on wiki update

**Comment by Mysak0CZ** [comment]
Done, info about this change is now available at [Hacking](https://tgstation13.org/wiki/Hacking#64px.7Clink.3DAirlock_64px.7Clink.3DAirlock_64px.7Clink.3DAirlock_Airlocks)

**Comment by kevinz000** [comment]
This is awful, it's been added to every command, security, AI, vault, and high security airlock. It's added atleast 5 seconds to the amount of time you need to take even with hotkeys to hack a door open, and now you can't even break open more than about 4 fucking doors without risking eye damage. Why? 
Also, you said "I will not add it to maps (if you don't tell me to do so), it is feature and unless someone uses metal or plasteel on doors with open panel, it will be like it is not in here"
Well, seems to me you just made door hacking, low pop rounds, AI less rounds, COUNTERING ROGUE AIs, and sometimes, doing your job a huge pain in the ass.
:-1:, why was this accepted?

**Comment by Core0verload** [comment]
I'm making a partial revert in next 24h.

**Comment by jughu** [comment]
This was secretlyba rcd buff anyways

**Comment by PKPenguin321** [comment]
Just remove welder requirement

**Comment by Cyberboss** [comment]
> Implying you should be able to easily hack those types of doors

Put a damn welding helmet on

**Comment by GunHog** [comment]
This buff needed to happen to offset the nerf to airlocks as a part of the obj damaging system. They should not be a joke of a barrier!

**Comment by PKPenguin321** [comment]
>Put a damn welding helmet on, you can find them where the welders are

Not true

**Comment by kevinz000** [comment]
Gee, airlocks worked fine in terms of security/strength for the last 3-4 years, and so did command channel. Why fix something that isn't broken?!

**Comment by optimumtact** [comment]
They weren't supposed to be on by default

**Comment by Mysak0CZ** [comment]
Sorry, i must have uploaded wrong commit, it was supposed to be only on vault and centcomm (max securoty) and on secure doors (basic security).
Wrong commit?: I have my server to play with friends, and we want to have it on command and security doors too.
